node/src/components/linear_chain/state.rs

```diff
@@ -410,6 +413,91 @@ mod tests {
+    }
+
+    #[test]
+    fn new_block_unvalidated_pending_sigs() {
+        let mut rng = TestRng::new();
+        let protocol_version = ProtocolVersion::V1_0_0;
+        let mut lc = LinearChain::new(protocol_version, 1u64, 1u64);
+        let block = Block::random(&mut rng);
+        let block_hash = *block.hash();
+        let block_era = block.header().era_id();
+        // Store some pending finality signatures
+        let _sig_a = add_pending(&mut lc, block_hash, block_era, true);
+        let _sig_b = add_pending(&mut lc, block_hash, block_era, false);
+        let _sig_c = add_pending(&mut lc, block_hash, block_era, false);
+
+        let execution_results = HashMap::new();
+        let outcomes = lc.handle_new_block(Box::new(block), execution_results);
+        // None of the signatures' creators have been confirmed to be bonded yet.
+        // We should not gossip/store/announce any signatures yet.
+        assert!(matches!(
+            &*outcomes,
+            [Outcome::StoreBlock(_), Outcome::StoreExecutionResults(_, _)]
+        ));
+    }
+
+    #[test]
+    fn new_block_bonded_pending_sigs() {
+        let mut rng = TestRng::new();
+        let protocol_version = ProtocolVersion::V1_0_0;
+        let mut lc = LinearChain::new(protocol_version, 1u64, 1u64);
+        let block = Block::random(&mut rng);
+        let block_hash = *block.hash();
+        let block_era = block.header().era_id();
+        // Store some pending finality signatures
+        let sig_a = add_pending(&mut lc, block_hash, block_era, true);
+        let sig_b = add_pending(&mut lc, block_hash, block_era, false);
+        let _sig_c = add_pending(&mut lc, block_hash, block_era, false);
+        // Mark two of the creators as bonded.
+        mark_bonded(&mut lc, sig_a.clone());
+        mark_bonded(&mut lc, sig_b.clone());
+        let execution_results = HashMap::new();
+        let outcomes = lc.handle_new_block(Box::new(block.clone()), execution_results.clone());
+        // `sig_a` and `sig_b` are valid and created by bonded validators.
+        let expected_outcomes = {
+            let mut tmp = vec![];
+            let mut block_signatures = BlockSignatures::new(block_hash, block_era);
+            block_signatures.insert_proof(sig_a.public_key.clone(), sig_a.signature);
+            block_signatures.insert_proof(sig_b.public_key.clone(), sig_b.signature);
+            tmp.push(Outcome::StoreBlockSignatures(block_signatures));
+            tmp.push(Outcome::Gossip(Box::new(sig_a.clone()))); // Only `sig_a` was created locally and we don't "regossip" incoming signatures.
+            tmp.push(Outcome::AnnounceSignature(Box::new(sig_a)));
+            tmp.push(Outcome::AnnounceSignature(Box::new(sig_b)));
+            tmp.push(Outcome::StoreBlock(Box::new(block)));
+            tmp.push(Outcome::StoreExecutionResults(
+                block_hash,
+                execution_results,
+            ));
+            tmp
```

Comment:
What if we mark C as bonded now, after `handle_new_block` — when would the execution results be stored?

Reply:
Storing of execution results and block is independent of the status of the signature verification. It's just that in this case, we return more outcomes. You can see that in `new_block_unvalidated_pending_sigs` test we return those two effects anyway.

Reply:
Sorry, I didn't mean to say "execution results". :man_facepalming: 
I mean: If we mark C as bonded now, after `handle_new_block`, we should get another `StoreBlockSignatures` outcome, should we? Would be good to test that, too.

Reply:
Yeah, good idea. I will add that.

Reply:
Done in [`1c06473` (#1291)](https://github.com/CasperLabs/casper-node/pull/1291/commits/1c06473d6a3bf36fb7f8e36ddc2b53d9bc0144d2).  It did not work as intended initially so I had to change `LinearChain` code as well.